src/rgw/rgw_op.cc

```diff
@@ -3514,3 +3514,197 @@ void RGWHandler::put_op(RGWOp *op)
+    ldout(s->cct, 0) << "DSS INFO: Converting to copy request. s->object: "
+                     << (s->object).name << ". Copy source: " << copysource << dendl;
+    s->info.env->set("HTTP_X_JCS_COPY_SOURCE", copysource.c_str());
+    s->info.env->set("HTTP_X_JCS_METADATA_DIRECTIVE", "COPY");
+    s->copy_source = s->info.env->get("HTTP_X_JCS_COPY_SOURCE");
+    if (s->copy_source) {
+      ret = RGWCopyObj::parse_copy_location(s->copy_source, s->src_bucket_name, s->src_object);
+      if (!ret || (store->ctx()->_conf->fault_inj_rename_op_parse_fail)) {
+        ldout(s->cct, 0) << "DSS INFO: Rename op failed to parse copy location" << dendl;
+        s->err.ret = -ERR_RENAME_FAILED;
+        return;
+      }
+    }
+
+    /* Perform copy operation */
+    RGWCopyObj_ObjStore_S3* copy_op = new RGWCopyObj_ObjStore_S3;
+    perform_external_op(copy_op);
+
+    if ((store->ctx()->_conf->fault_inj_rename_op_sleep_after_copy)) {
+        // Fault injection to test atomicity
+        sleep(60);
+    }
+
+    if ((s->err.http_ret != 200) ||
+        (s->err.ret != 0)) {
+        ldout(s->cct, 0) << "DSS ERROR: Copy object failed during rename op."
+                         << " . Return status: " << s->err.ret
+                         << " . Return HTTP code: " << s->err.http_ret
+                         << " . Return Message: " << copy_op->get_request_state()->err.message
+                         << dendl;
+        s->err.ret = -ERR_RENAME_FAILED;
+        return;
+    }
+    ldout(s->cct, 0) << "DSS INFO: Rename op: copy done" << dendl;
+
+    /* Tweek the request for a delete obj operation and perform delete op */
+    new_obj.dss_duplicate(&(s->object));
+    (s->object).dss_duplicate(&orig_object);
+    RGWDeleteObj_ObjStore_S3* del_op = new RGWDeleteObj_ObjStore_S3;
+    delete_rgw_object(del_op);
+    if ((s->err.http_ret != 200) ||
+        (s->err.ret != 0)) {
+        ldout(s->cct, 0) << "DSS ERROR: Delete object failed during rename op."
+                         << " . Return status: " << s->err.ret
+                         << " . Return HTTP code: " << s->err.http_ret
+                         << " . Return Message: " << del_op->get_request_state()->err.message
+                         << dendl;
+
+        /* Revert the copy op */
+        int ret_orig, ret_newobj;
+        ret_orig = check_obj(s->object);
+        ret_newobj = check_obj(new_obj);
+        if (ret_orig < 0) {
+            // Delete failed but we don't have original object!!
+            if (ret_newobj < 0) {
+                // We are in a soup. Data lost. This is not the case we will ever end up in.
+                s->err.ret = -ERR_RENAME_DATA_LOST;
+            } else {
+                // Everything normal
+                if (!store->ctx()->_conf->fault_inj_rename_op_delete_fail) {
+                    s->err.http_ret = 200;
+                    s->err.ret = 0;
+                }
+            }
+        } else {
+            if (ret_newobj >= 0) {
```

Comment:
Yes.. this is just added protection in the unlikely/impossible scenario that copy fails without throwing an error.